contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,570 @@
+
+    /// Version of the initialized contract data
+    /// NB: Contract versioning starts from 1.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to
+    /// - 0 right after deployment when no initializer is invoked yet
+    /// - N after calling initialize() during deployment from scratch, where N is the current contract version
+    /// - N after upgrading contract from the previous version (after calling finalize_vN())
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.StakingRouter.contractVersion");
+
+    /// @dev Credentials which allows the DAO to withdraw Ether on the 2.0 side
+    bytes32 internal constant WITHDRAWAL_CREDENTIALS_POSITION = keccak256("lido.StakingRouter.withdrawalCredentials");
+
+    uint256 public constant DEPOSIT_SIZE = 32 ether;
+
+    uint256 public constant TOTAL_BASIS_POINTS = 10000;
+
+    /// @dev total count of staking modules
+    uint256 private _stakingModulesCount;
+
+    /// @dev id of the last added staking module. This counter grow on staking modules adding
+    uint24 private _lastStakingModuleId;
+
+    /// @dev mapping is used instead of array to allow to extend the StakingModule
+    mapping(uint256 => StakingModule) private _stakingModules;
+
+    /// @dev Position of the staking modules in the `_stakingModules` map, plus 1 because
+    ///      index 0 means a value is not in the set.
+    mapping(uint24 => uint256) private _stakingModuleIndicesOneBased;
+
+    constructor(address _depositContract) BeaconChainDepositor(_depositContract) {}
+
+    function initialize(address _admin, bytes32 _withdrawalCredentials) external {
+        if (_admin == address(0)) revert ErrorZeroAddress("_admin");
+        if (CONTRACT_VERSION_POSITION.getStorageUint256() != 0) revert ErrorBaseVersion();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+
+        CONTRACT_VERSION_POSITION.setStorageUint256(1);
+        WITHDRAWAL_CREDENTIALS_POSITION.setStorageBytes32(_withdrawalCredentials);
+
+        emit ContractVersionSet(1);
+        emit WithdrawalCredentialsSet(_withdrawalCredentials);
+    }
+
+    /// @dev prohibit direct transfer to contract
+    receive() external payable {
+        revert ErrorDirectETHTransfer();
+    }
+
+    /**
+     * @notice register a new module
+     * @param _name name of module
+     * @param _stakingModuleAddress target percent of total keys in protocol, in BP
+     * @param _targetShare target total stake share
+     * @param _moduleFee fee of the module taken from the consensus layer rewards
+     * @param _treasuryFee treasury fee
+     */
+    function addModule(
+        string memory _name,
+        address _stakingModuleAddress,
+        uint16 _targetShare,
+        uint16 _moduleFee,
+        uint16 _treasuryFee
+    ) external onlyRole(MODULE_MANAGE_ROLE) {
+        if (_targetShare > TOTAL_BASIS_POINTS) revert ErrorValueOver100Percent("_targetShare");
```

Comment:
add check that `_moduleFee ` + `_treasuryFee` < 100%

Reply:
added